src/main/java/com/es/phoneshop/controller/servlet/ProductDetailsPageServlet.java

```diff
@@ -1,31 +1,94 @@
+    private RecentlyViewedService recentlyViewedService = RecentlyViewedService.getInstance();
 
     @Override
-    protected void doGet(HttpServletRequest request, HttpServletResponse response) throws ServletException, IOException {
-        String productId = request.getPathInfo().substring(1);
+    protected void doGet(HttpServletRequest req, HttpServletResponse resp) throws ServletException, IOException {
+        Optional<Product> product = extractProductOrSendError(req, resp);
+        if (product.isEmpty()) {
+            return;
+        }
+        req.setAttribute(REQUEST_ATTRIBUTE_PRODUCT, product.get());
+        req.getRequestDispatcher(PRODUCT_DETAILS_PATH).forward(req, resp);
+        recentlyViewedService.updateRecentlyViewedLine(req.getSession(), product.get());
+    }
+
+    @Override
+    protected void doPost(HttpServletRequest req, HttpServletResponse resp) throws ServletException, IOException {
+        Optional<Product> product = extractProductOrSendError(req, resp);
+        Optional<Integer> quantity = extractQuantityOrForward(req, resp);
+        if (product.isEmpty() || quantity.isEmpty()) {
+            return;
+        }
+        addProductToCartOrForward(req, resp, product.get(), quantity.get());
+    }
+
+    private Optional<Product> extractProductOrSendError(HttpServletRequest req, HttpServletResponse resp)
+            throws IOException {
+        String productId = req.getPathInfo().substring(1);
         try {
-            request.setAttribute(PRODUCT_CODE_ATTRIBUTE, arrayListProductDao.findProduct(Long.valueOf(productId)));
-            request.getRequestDispatcher(PRODUCT_DETAILS_PATH).forward(request, response);
+            return Optional.of(dao.findProduct(Long.valueOf(productId)));
         } catch (NumberFormatException | ItemNotFoundException e) {
-            request.setAttribute(ERROR_MESSAGE_ATTRIBUTE, String.format(ERROR_MESSAGE_PART, productId));
-            response.sendError(NOT_FOUND_ERROR_CODE);
+            req.setAttribute(REQUEST_ATTRIBUTE_MESSAGE, String.format(ERROR_NOT_FOUND, productId));
+            resp.sendError(NOT_FOUND_ERROR_CODE);
+            return Optional.empty();
+        }
+    }
+
+    private Optional<Integer> extractQuantityOrForward(HttpServletRequest req, HttpServletResponse resp)
+            throws ServletException, IOException {
+        try {
+            return Optional.of(NumberFormat.getInstance(
+                    req.getLocale()).parse(req.getParameter(REQUEST_PARAM_QUANTITY)).intValue());
+        } catch (NumberFormatException | ParseException e) {
+            req.setAttribute(REQUEST_ATTRIBUTE_ADDING_TO_CART_MESSAGE, false);
+            req.setAttribute(REQUEST_ATTRIBUTE_ERROR, ERROR_NOT_NUMBER);
+            doGet(req, resp);
+            return Optional.empty();
+        }
+    }
+
+    private void addProductToCartOrForward
+            (HttpServletRequest req, HttpServletResponse resp, Product product, int quantity)
```

Comment:
Minor point: it's commonly used to wrap long method-declaration lines in the way like:
```
private void addProductToCartOrForward(
         HttpServletRequest req, HttpServletResponse resp, Product product, int quantity)
```
or
```
private void addProductToCartOrForward(HttpServletRequest req, HttpServletResponse resp,
                                       Product product, int quantity)
```